### PR TITLE
Implement support of date/time format specifiers in log and error log file names

### PR DESCRIPTION
**Pull request by sirvickr**
### Changelog category (leave one):

- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Added the possibility to use date and time format specifiers in log and error log file names, either in config files (`log` and `errorlog` tags) or command line arguments (`--log-file` and `--errorlog-file`)

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
This is an automated comment for commit 5b102ce7d44d678a674d29e4140a60950c69f537 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51945/5b102ce7d44d678a674d29e4140a60950c69f537/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

**Comment by rvasin** [comment]
Hi, could anyone be assigned to this PR? So Victor could start working on reviewers suggestions.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
LGTM.

The `localtime_r` and `stringstream` usage was suspicious to me while reviewing, but it looks like we cannot support all these format substitutions easier, and the method will be called very rarely, and it is only static (buried deep enough).

File: src/Loggers/Loggers.cpp [hunk 1]
```diff
@@ -34,6 +34,16 @@ static std::string createDirectory(const std::string & file)
...
 }
 
+static std::string renderFileNameTemplate(time_t now, const std::string & file_path)
+{
+    fs::path path{file_path};
+    std::tm buf;
+    localtime_r(&now, &buf);
+    std::ostringstream ss; // STYLE_CHECK_ALLOW_STD_STRING_STREAM
+    ss << std::put_time(&buf, file_path.c_str());
+    return path.replace_filename(ss.str());
```
Comment (amosbird) [inline src/Loggers/Loggers.cpp:44]: This breaks relative path configuration for non-daemon servers. Consider the following example:

```
    <logger>
        <level>trace</level>
        <log>logs/clickhouse-server.log</log>
        <errorlog>logs/clickhouse-server.err.log</errorlog>
        <size>1000M</size>
        <count>10</count>
    </logger>
```

The log path will be modified as `logs/logs/clickhouse-server.log`.

Comment (sirvickr) [reply src/Loggers/Loggers.cpp:44]: That's the exact reason why those format specifiers currently do not apply to the "path" part - only to the file name.

Comment (sirvickr) [reply src/Loggers/Loggers.cpp:44]: @amosbird, `path.replace_filename()` does the trick.
If someone decides another day to apply the specifiers to the full path, they will have to take into consideration this issue about non-daemon servers.

Comment (amosbird) [reply src/Loggers/Loggers.cpp:44]: > That's the exact reason why those format specifiers currently do not apply to the "path" part - only to the file name.

I'm confused. If that's the expected behavior, why not do the following:

```
    ss << std::put_time(&buf, path.filename().c_str());
```

Comment (sirvickr) [reply src/Loggers/Loggers.cpp:44]: @amosbird, you are right, my code looks like an ugly typo, to say the least. This one is more concise, clear and correct:
```cpp
ss << std::put_time(&buf, path.filename().c_str());
```
It looks like it should be amended.

Comment (sirvickr) [reply src/Loggers/Loggers.cpp:44]: @amosbird, you are right, my code looks like an ugly typo, to say the least. This one is more concise, clear and correct:
```cpp
ss << std::put_time(&buf, path.filename().c_str());
```
It looks like it should be amended.

Comment (sirvickr) [reply src/Loggers/Loggers.cpp:44]: I guess, `replace_filename()` extracts the file name from its argument, otherwise it wouldn't work properly.

Comment (sirvickr) [reply src/Loggers/Loggers.cpp:44]: And thank you for noticing.
I'm thinking of creating a "follow-up" PR to clean up this line of code. Or may be you'd rather amend this code by yourself?

Comment (amosbird) [reply src/Loggers/Loggers.cpp:44]: > And thank you for noticing.

No problem! 

> I'm thinking of creating a "follow-up" PR to clean up this line of code. 

Perfect. Please go ahead for amending it :)

**Pull request merged**